lightning/src/routing/router.rs

```diff
@@ -343,8 +347,8 @@ fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
 /// The fees on channels from us to next-hops are ignored (as they are assumed to all be
 /// equal), however the enabled/disabled bit on such channels as well as the
 /// htlc_minimum_msat/htlc_maximum_msat *are* checked as they may change based on the receiving node.
-pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, payee: &PublicKey, payee_features: Option<InvoiceFeatures>, first_hops: Option<&[&ChannelDetails]>,
-	last_hops: &[&RouteHintHop], final_value_msat: u64, final_cltv: u32, logger: L) -> Result<Route, LightningError> where L::Target: Logger {
+pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, payee: &PublicKey, payee_features: Option<InvoiceFeatures>, first_hops: Option<&[ChannelDetails]>,
```

Comment:
Oops, ugh, the `ChannelDetails` reference is here mostly for the bindings - the bindings take an array of pointers with some additional metadata, so they can't trivially convert them to `[ChannelDetails]` without a bunch of copies.

Reply:
Dropped the commit but it does make the user do more work at the call site. Here is what the guide is looking like:

```rust
let amt_pico_btc = invoice.amount_pico_btc()
    .expect("ERROR: invalid invoice: must contain amount to pay");
let amt_msat = amt_pico_btc / 10;
let payer_pubkey = channel_manager.get_our_node_id();
let network_graph = router.network_graph.read().unwrap();
let payee_pubkey = invoice.recover_payee_pub_key();
let payee_features = invoice.features().cloned();
let first_hops = channel_manager.list_usable_channels();
let last_hops = invoice.route_hints();
let final_cltv = invoice.min_final_cltv_expiry() as u32;

let route = router::get_route(
    &payer_pubkey, &network_graph, &payee_pubkey, payee_features,
    Some(&first_hops.iter().collect::<Vec<_>>()), &last_hops,
    amt_msat, final_cltv, logger.clone(),
).expect("ERROR: failed to find route");

let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
let payment_secret = invoice.payment_secret().cloned();

channel_manager.send_payment(&route, payment_hash, &payment_secret)
    .expect("ERROR: failed to send payment");
```

Later, we may want take a pass through how our interfaces to see how well they mesh together.

Reply:
Yea, its very not-ideal, I agree. I'm not really sure how to fix it in the bindings context, though, sadly.